### PR TITLE
 Testing volume taint policy

### PR DESCRIPTION
**Pull request by Swarnalatha-K**
Signed-off-by: swarna <swarnalatha@cloudbyte.com>

**What this PR does / why we need it**:
  1.Verifying the replica deployment on tainted nodes.
  2. verifying the storage pool path.
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
@ksatchit please review this PR.
This test will fail in CI untill the volume taint policy issue is fixed.

### CODE REVIEW

**Comment by Swarnalatha-K** [comment]
In volume policy replicas are going to pending state so the script fail while checking replica status.
Output:

TASK [Check whether maya-apiserver pod is running] *********************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/node1.yml:10
FAILED - RETRYING: Check whether maya-apiserver pod is running (2 retries left).
FAILED - RETRYING: Check whether maya-apiserver pod is running (1 retries left).
fatal: [localhost -> None]: FAILED! => {"attempts": 2, "changed": true, "cmd": "source ~/.profile; kubectl get pods --all-namespaces | grep maya-apiserver", "delta": "0:00:01.224607", "end": "2018-04-09 07:09:15.490469", "msg": "non-zero return code", "rc": 1, "start": "2018-04-09 07:09:14.265862", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
...ignoring

TASK [Delete the openebs operator] *************************************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/node1.yml:23
skipping: [localhost] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Confirm pod has been deleted] ************************************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/node1.yml:29
skipping: [localhost] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Taint the nodes for maya-apiserver] ******************************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/node1.yml:42
changed: [localhost -> None] => {"changed": true, "cmd": "source ~/.profile; kubectl taint nodes kubeminion-01 key1=value1:NoSchedule", "delta": "0:00:00.931451", "end": "2018-04-09 07:09:17.547204", "rc": 0, "start": "2018-04-09 07:09:16.615753", "stderr": "", "stderr_lines": [], "stdout": "node \"kubeminion-01\" tainted", "stdout_lines": ["node \"kubeminion-01\" tainted"]}

TASK [Taint the nodes for the pvc] *************************************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/node1.yml:48
changed: [localhost -> None] => {"changed": true, "cmd": "source ~/.profile; kubectl taint nodes kubeminion-02 key2=value2:NoSchedule; kubectl taint nodes kubeminion-03 key2=value2:NoSchedule", "delta": "0:00:02.416097", "end": "2018-04-09 07:09:21.041757", "rc": 0, "start": "2018-04-09 07:09:18.625660", "stderr": "", "stderr_lines": [], "stdout": "node \"kubeminion-02\" tainted\nnode \"kubeminion-03\" tainted", "stdout_lines": ["node \"kubeminion-02\" tainted", "node \"kubeminion-03\" tainted"]}

TASK [Check nodes are tainted] *****************************************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/node1.yml:54
changed: [localhost -> None] => {"attempts": 1, "changed": true, "cmd": "source ~/.profile; kubectl describe node kubeminion-01 kubeminion-02 kubeminion-03 | grep NoSchedule", "delta": "0:00:01.611186", "end": "2018-04-09 07:09:24.270090", "rc": 0, "start": "2018-04-09 07:09:22.658904", "stderr": "", "stderr_lines": [], "stdout": "Taints:             key1=value1:NoSchedule\nTaints:             key2=value2:NoSchedule\nTaints:             key2=value2:NoSchedule", "stdout_lines": ["Taints:             key1=value1:NoSchedule", "Taints:             key2=value2:NoSchedule", "Taints:             key2=value2:NoSchedule"]}

TASK [Get $HOME of K8s master for kubernetes user] *********************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/node1.yml:65
changed: [localhost -> None] => {"changed": true, "cmd": "source ~/.profile; echo $HOME", "delta": "0:00:00.006681", "end": "2018-04-09 07:09:25.124833", "rc": 0, "start": "2018-04-09 07:09:25.118152", "stderr": "", "stderr_lines": [], "stdout": "/home/ubuntu", "stdout_lines": ["/home/ubuntu"]}

TASK [Copy the openebs-operator to kube master] ************************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/node1.yml:72
changed: [localhost -> None] => {"changed": true, "checksum": "5ebe81f71efb2c6e839c78f900ef42987722ff77", "dest": "/home/ubuntu/openebs-operator-ns-openebs.yaml", "gid": 1000, "group": "ubuntu", "md5sum": "f82d37cecf529782416fc77e05000acd", "mode": "0664", "owner": "ubuntu", "size": 6927, "src": "/home/ubuntu/.ansible/tmp/ansible-tmp-1523257765.3-124060051757265/source", "state": "file", "uid": 1000}

TASK [Replacing the key in openebs-operator yaml] **********************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/node1.yml:78
changed: [localhost -> None] => {"changed": true, "msg": "2 replacements made"}

TASK [Replacing the value in openebs-operator yaml] ********************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/node1.yml:85
changed: [localhost -> None] => {"changed": true, "msg": "2 replacements made"}

TASK [deploy the openebs-operator with openebs namespace] **************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/node1.yml:92
changed: [localhost -> None] => {"changed": true, "cmd": "source ~/.profile; kubectl apply -f \"openebs-operator-ns-openebs.yaml\"", "delta": "0:00:03.240872", "end": "2018-04-09 07:09:36.096648", "rc": 0, "start": "2018-04-09 07:09:32.855776", "stderr": "", "stderr_lines": [], "stdout": "namespace \"openebs\" created\nserviceaccount \"openebs-maya-operator\" created\nclusterrole \"openebs-maya-operator\" created\nclusterrolebinding \"openebs-maya-operator\" created\ndeployment \"maya-apiserver\" created\nservice \"maya-apiserver-service\" created\ndeployment \"openebs-provisioner\" created\ncustomresourcedefinition \"storagepoolclaims.openebs.io\" created\ncustomresourcedefinition \"storagepools.openebs.io\" created\nstorageclass \"openebs-standard\" created\ncustomresourcedefinition \"volumepolicies.openebs.io\" created", "stdout_lines": ["namespace \"openebs\" created", "serviceaccount \"openebs-maya-operator\" created", "clusterrole \"openebs-maya-operator\" created", "clusterrolebinding \"openebs-maya-operator\" created", "deployment \"maya-apiserver\" created", "service \"maya-apiserver-service\" created", "deployment \"openebs-provisioner\" created", "customresourcedefinition \"storagepoolclaims.openebs.io\" created", "customresourcedefinition \"storagepools.openebs.io\" created", "storageclass \"openebs-standard\" created", "customresourcedefinition \"volumepolicies.openebs.io\" created"]}

TASK [Check whether maya-apiserver pod is deployed] ********************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/node1.yml:98
FAILED - RETRYING: Check whether maya-apiserver pod is deployed (5 retries left).
changed: [localhost -> None] => {"attempts": 2, "changed": true, "cmd": "source ~/.profile; kubectl get pods --all-namespaces | grep maya-apiserver", "delta": "0:00:01.235341", "end": "2018-04-09 07:11:43.864992", "rc": 0, "start": "2018-04-09 07:11:42.629651", "stderr": "", "stderr_lines": [], "stdout": "openebs       maya-apiserver-5f6f5fcd7-jpxg8          1/1       Running   0          2m", "stdout_lines": ["openebs       maya-apiserver-5f6f5fcd7-jpxg8          1/1       Running   0          2m"]}

TASK [Copy the volume-policies yaml to kube master] ********************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/node1.yml:108
changed: [localhost -> None] => {"changed": true, "checksum": "85781cb1c26cf8ccab48f89463ce3d8d36527f3e", "dest": "/home/ubuntu/vol-policies.yaml", "gid": 1000, "group": "ubuntu", "md5sum": "da8bc9448bbabf4fa07f9a071b92955e", "mode": "0664", "owner": "ubuntu", "size": 9158, "src": "/home/ubuntu/.ansible/tmp/ansible-tmp-1523257904.26-26640599592839/source", "state": "file", "uid": 1000}

TASK [Replacing the key in volume-policies yaml] ***********************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/node1.yml:114
changed: [localhost -> None] => {"changed": true, "msg": "1 replacements made"}

TASK [Replacing the value in volume-policies yaml] *********************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/node1.yml:121
changed: [localhost -> None] => {"changed": true, "msg": "1 replacements made"}

TASK [deploy the volume-policies yaml] *********************************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/node1.yml:128
changed: [localhost -> None] => {"changed": true, "cmd": "source ~/.profile; kubectl apply -f \"vol-policies.yaml\"", "delta": "0:00:02.550750", "end": "2018-04-09 07:11:52.539235", "rc": 0, "start": "2018-04-09 07:11:49.988485", "stderr": "", "stderr_lines": [], "stdout": "storagepool \"sp-mntdir\" created\nstorageclass \"openebs-taints-0.6.0\" created\nvolumepolicy \"openebs-policy-taints-0.6.0\" created\nconfigmap \"volume-service-0.6.0\" created\nconfigmap \"volume-path-0.6.0\" created\nconfigmap \"volume-controller-0.6.0\" created\nconfigmap \"volume-replica-0.6.0\" created", "stdout_lines": ["storagepool \"sp-mntdir\" created", "storageclass \"openebs-taints-0.6.0\" created", "volumepolicy \"openebs-policy-taints-0.6.0\" created", "configmap \"volume-service-0.6.0\" created", "configmap \"volume-path-0.6.0\" created", "configmap \"volume-controller-0.6.0\" created", "configmap \"volume-replica-0.6.0\" created"]}

TASK [download test-pvc file] ******************************************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/node1.yml:135
changed: [localhost -> None] => {"attempts": 1, "changed": true, "checksum_dest": "ce24440083e048cc52f949741a867c431a15734c", "checksum_src": "f94bef95544501005c9fda43342beb00bf87bbf4", "dest": "/home/ubuntu/test-pvc.yaml", "gid": 1000, "group": "ubuntu", "md5sum": "7ef16cd931722d2f35c929e4bf42f26d", "mode": "0664", "msg": "OK (189 bytes)", "owner": "ubuntu", "size": 189, "src": "/tmp/tmpdb4Q7b", "state": "file", "status_code": 200, "uid": 1000, "url": "https://raw.githubusercontent.com/openebs/openebs/master/e2e/ansible/playbooks/hyperconverged/test-k8s-storage-pool/test-pvc.yaml"}

TASK [Replacing the storage class name in test-pvc yaml] ***************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/node1.yml:146
changed: [localhost -> None] => {"changed": true, "msg": "1 replacements made"}

TASK [Create  namespace for PVC] ***************************************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/node1.yml:153
changed: [localhost -> None] => {"changed": true, "cmd": "kubectl create ns \"percona\"", "delta": "0:00:01.337980", "end": "2018-04-09 07:11:59.597137", "rc": 0, "start": "2018-04-09 07:11:58.259157", "stderr": "", "stderr_lines": [], "stdout": "namespace \"percona\" created", "stdout_lines": ["namespace \"percona\" created"]}

TASK [deploy the test-pvc yaml] ****************************************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/node1.yml:161
changed: [localhost -> None] => {"changed": true, "cmd": "source ~/.profile; kubectl apply -f test-pvc.yaml -n \"percona\"", "delta": "0:00:01.515099", "end": "2018-04-09 07:12:02.563372", "rc": 0, "start": "2018-04-09 07:12:01.048273", "stderr": "", "stderr_lines": [], "stdout": "persistentvolumeclaim \"vut\" created", "stdout_lines": ["persistentvolumeclaim \"vut\" created"]}

TASK [Confirm volume container is running] *****************************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/node1.yml:167
FAILED - RETRYING: Confirm volume container is running (5 retries left).
FAILED - RETRYING: Confirm volume container is running (4 retries left).
FAILED - RETRYING: Confirm volume container is running (3 retries left).
FAILED - RETRYING: Confirm volume container is running (2 retries left).
FAILED - RETRYING: Confirm volume container is running (1 retries left).
failed: [localhost -> None] (item=ctrl) => {"attempts": 5, "changed": true, "cmd": "source ~/.profile; kubectl get pods --all-namespaces | grep pvc | grep ctrl | grep Running | wc -l", "delta": "0:00:01.516425", "end": "2018-04-09 07:14:46.648930", "item": "ctrl", "rc": 0, "start": "2018-04-09 07:14:45.132505", "stderr": "", "stderr_lines": [], "stdout": "0", "stdout_lines": ["0"]}
FAILED - RETRYING: Confirm volume container is running (5 retries left).
FAILED - RETRYING: Confirm volume container is running (4 retries left).
FAILED - RETRYING: Confirm volume container is running (3 retries left).
FAILED - RETRYING: Confirm volume container is running (2 retries left).
FAILED - RETRYING: Confirm volume container is running (1 retries left).
failed: [localhost -> None] (item=rep) => {"attempts": 5, "changed": true, "cmd": "source ~/.profile; kubectl get pods --all-namespaces | grep pvc | grep rep | grep Running | wc -l", "delta": "0:00:00.738904", "end": "2018-04-09 07:17:28.975850", "item": "rep", "rc": 0, "start": "2018-04-09 07:17:28.236946", "stderr": "", "stderr_lines": [], "stdout": "0", "stdout_lines": ["0"]}

File: e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/test-volume-taints.yml [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+         shell: source ~/.profile; kubectl taint nodes {{ node1 }} {{ operator_key }}={{ operator_value }}:NoSchedule
+         args:
+           executable: /bin/bash
+         delegate_to: "{{groups['kubernetes-kubemasters'].0}}"
+
+       - name: Taint the nodes for the pvc
+         shell: source ~/.profile; kubectl taint nodes {{ node2 }} {{ pvc_key }}={{ pvc_value }}:NoSchedule; kubectl taint nodes {{ node3 }} {{ pvc_key }}={{ pvc_value }}:NoSchedule
+         args:
+           executable: /bin/bash
+         delegate_to: "{{groups['kubernetes-kubemasters'].0}}"
```
Comment (ksatchit) [inline e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/test-volume-taints.yml]: Please use  ansible loops to iterate over the node & taint lists

File: e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/test-volume-taints.yml [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+           regexp: 'key: jk'
+           replace: 'key: {{ pvc_key }}'
+         delegate_to: "{{ groups['kubernetes-kubemasters'].0 }}"
+
+       - name: Replacing the value in volume-policies yaml
+         replace:
+           path: vol-policies.yaml
+           regexp: 'value: jv'
+           replace: 'key: {{ pvc_value }}'
+         delegate_to: "{{ groups['kubernetes-kubemasters'].0 }}"
```
Comment (ksatchit) [inline e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/test-volume-taints.yml]: The regex_replace steps can be done in a single step

File: e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/test-volume-taints.yml [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+         set_fact:
+           rep1_node: "{{ result.stdout.split()[7] }}"
+
+       - name: Set replica pod name to variable
+         set_fact:
+           rep2_node: "{{ result.stdout.split()[15] }}"
+
+       - name: Verifying the replicas are deployed on tainted nodes
+         shell: echo " replicas are running on the tainted nodes"
+         when:  rep1_node  == node1 or rep1_node == node2 and rep2_node == node1 or rep2_node == node2
```
Comment (ksatchit) [inline e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/test-volume-taints.yml]: Please use ansible debug module with failure conditions

File: e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/test-volume-taints.yml [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+         delay: 30
+         retries: 5
+         with_items:
+           - ctrl
+           - rep
+         delegate_to: "{{ groups['kubernetes-kubemasters'].0 }}"
+         
+ 
+       - name:
+         shell: source ~/.profile; kubectl get pods --all-namespaces -o wide | grep rep
```
Comment (ksatchit) [inline e2e/ansible/playbooks/hyperconverged/test-k8s-volume-taints/test-volume-taints.yml]: Please use the volume replica deployment name to obtain the list of replicas and their nodes

**Pull request closed without merging**